<a href="https://colab.research.google.com/github/Ghouyd/AI-Phishing-Website-Detection/blob/main/AI_Assignment_Phishing_URL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Phishing Website Detection using Random Forest

## Step 1: Import Libraries

In [1]:
import pandas as pd

## Step 2: Extract the Dataset

In [3]:
import zipfile

with zipfile.ZipFile("archive.zip", "r") as zip_ref:
    zip_ref.extractall()

## Step 3: Load the Dataset

In [4]:
data = pd.read_csv("dataset.csv")
data.head()

,index,having_IPhaving_IP_Address,URLURL_Length,Shortining_Service,having_At_Symbol,double_slash_redirecting,Prefix_Suffix,having_Sub_Domain,SSLfinal_State,Domain_registeration_length,...,popUpWidnow,Iframe,age_of_domain,DNSRecord,web_traffic,Page_Rank,Google_Index,Links_pointing_to_page,Statistical_report,Result
0,1,-1,1,1,1,-1,-1,-1,-1,-1,...,1,1,-1,-1,-1,-1,1,1,-1,-1
1,2,1,1,1,1,1,-1,0,1,-1,...,1,1,-1,-1,0,-1,1,1,1,-1
2,3,1,0,1,1,1,-1,-1,-1,-1,...,1,1,1,-1,1,-1,1,0,-1,-1
3,4,1,0,1,1,1,-1,-1,-1,1,...,1,1,-1,-1,1,-1,1,-1,1,-1
4,5,1,0,-1,1,1,-1,1,1,-1,...,-1,1,-1,-1,0,-1,1,1,1,1


## Step 4: Explore the Dataset

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11055 entries, 0 to 11054
Data columns (total 32 columns):
 #   Column                       Non-Null Count  Dtype
---  ------                       --------------  -----
 0   index                        11055 non-null  int64
 1   having_IPhaving_IP_Address   11055 non-null  int64
 2   URLURL_Length                11055 non-null  int64
 3   Shortining_Service           11055 non-null  int64
 4   having_At_Symbol             11055 non-null  int64
 5   double_slash_redirecting     11055 non-null  int64
 6   Prefix_Suffix                11055 non-null  int64
 7   having_Sub_Domain            11055 non-null  int64
 8   SSLfinal_State               11055 non-null  int64
 9   Domain_registeration_length  11055 non-null  int64
 10  Favicon                      11055 non-null  int64
 11  port                         11055 non-null  int64
 12  HTTPS_token                  11055 non-null  int64
 13  Request_URL                  11055 non-null  i

## Step 5: Define Features and Target

>



In [6]:
X = data.drop("Result", axis=1)
y = data["Result"]

## Step 6: Split the Dataset

In [9]:
from sklearn.model_selection import train_test_split

X = X.drop("index", axis=1)

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.4,
    random_state=42
)

In [10]:
X.shape

(11055, 30)

## Step 7: Build the Random Forest Model

In [12]:
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier()

## Step 8: Train the Model

In [13]:
model.fit(X_train, y_train)

RandomForestClassifier()

## Step 9: Make Predictions

In [14]:
y_pred = model.predict(X_test)

## Step 10: Evaluate the Model

In [16]:
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)

print("Accuracy:", round(accuracy * 100, 2), "%")

Accuracy: 96.7 %


## Step 11: Generate the Classification Report

In [17]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

          -1       0.97      0.95      0.96      1906
           1       0.96      0.98      0.97      2516

    accuracy                           0.97      4422
   macro avg       0.97      0.96      0.97      4422
weighted avg       0.97      0.97      0.97      4422



## Step 12: Confusion Matrix

In [22]:
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, y_pred)

print("Confusion Matrix:")
print(cm)

Confusion Matrix:
[[1810   96]
 [  50 2466]]


## Step 13: Evaluate Training and Testing Performance

In [21]:
y_train_pred = model.predict(X_train)

train_accuracy = accuracy_score(y_train, y_train_pred)

print(f"Training Accuracy: {train_accuracy * 100:.2f}%")
print(f"Testing Accuracy: {accuracy * 100:.2f}%")

Training Accuracy: 99.07%
Testing Accuracy: 96.70%


## Step 14: Conclusion
The Random Forest model achieved a testing accuracy of 96.70%, indicating strong performance in detecting phishing websites. The training and testing accuracies are close (99.07% vs. 96.70%), suggesting that the model generalizes well and does not show significant overfitting or underfitting.